packages/app/src/firebaseServerApp.ts

```diff
@@ -25,6 +25,77 @@ import { deleteApp } from './api';
 import { ComponentContainer } from '@firebase/component';
 import { FirebaseAppImpl } from './firebaseApp';
 import { ERROR_FACTORY, AppError } from './errors';
+import type { KeyObject } from 'node:crypto';
+
+let memoizedPublicKeys: Map<string, CryptoKey | KeyObject> | undefined;
+let memoizedPublicKeysExpireAt: Date | undefined;
+
+async function getPublicKeys(): Promise<Map<string, CryptoKey | KeyObject>> {
+  if (
+    memoizedPublicKeys &&
+    memoizedPublicKeysExpireAt &&
+    memoizedPublicKeysExpireAt < new Date()
+  ) {
+    return memoizedPublicKeys;
+  }
+  const response = await fetch(
+    'https://www.googleapis.com/robot/v1/metadata/jwk/securetoken@system.gserviceaccount.com'
+  );
+  if (!response.ok) {
+    return memoizedPublicKeys || Promise.reject();
+  }
+  const certificates = await response.json().catch(() => undefined);
+  if (!certificates) {
+    return memoizedPublicKeys || Promise.reject();
+  }
+  const cacheControl = response.headers.get('Cache-Control');
+  const maxAge = +(cacheControl?.match(/max-age=(\d+)/)?.[1] ?? 0);
+  memoizedPublicKeysExpireAt = new Date(+new Date() + maxAge * 1_000);
+  memoizedPublicKeys = new Map<string, CryptoKey | KeyObject>();
+  for (const jwk of certificates.keys as Array<{
+    kid: string;
+    [key: string]: string;
+  }>) {
+    const key =
+      typeof process !== 'undefined' && process.release?.name === 'node'
+        ? // eslint-disable-next-line @typescript-eslint/no-require-imports
+          require('node:crypto').createPublicKey({ key: jwk, format: 'jwk' })
+        : // @ts-expect-error
+          await crypto.subtle.importKey(
+            'jwk',
+            jwk,
+            {
+              name: 'RSASSA-PKCS1-v1_5',
+              hash: 'SHA-256'
+            },
+            false,
+            ['verify']
+          );
+    memoizedPublicKeys.set(jwk.kid, key);
+  }
+  return memoizedPublicKeys;
+}
+
+function base64decode(base64Contents: string): string {
```

Comment:
are there already utils for this ops? look into

Reply:
And if not, then add this to the util file.